sentry-android-replay/src/main/java/io/sentry/android/replay/ReplayIntegration.kt

```diff
@@ -0,0 +1,317 @@
+    private val isRecording = AtomicBoolean(false)
+    private val currentReplayId = AtomicReference<SentryId>()
+    private val segmentTimestamp = AtomicReference<Date>()
+    private val currentSegment = AtomicInteger(0)
+    private val saver =
+        Executors.newSingleThreadScheduledExecutor(ReplayExecutorServiceThreadFactory())
+
+    private val screenBounds by lazy(NONE) {
+        // PixelCopy takes screenshots including system bars, so we have to get the real size here
+        val wm = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
+        if (VERSION.SDK_INT >= VERSION_CODES.R) {
+            wm.currentWindowMetrics.bounds
+        } else {
+            val screenBounds = Point()
+            @Suppress("DEPRECATION")
+            wm.defaultDisplay.getRealSize(screenBounds)
+            Rect(0, 0, screenBounds.x, screenBounds.y)
+        }
+    }
+
+    private val aspectRatio by lazy(NONE) {
+        screenBounds.bottom.toFloat() / screenBounds.right.toFloat()
+    }
+
+    private val recorderConfig by lazy(NONE) {
+        ScreenshotRecorderConfig(
+            recordingWidth = (720 / aspectRatio).roundToInt(),
+            recordingHeight = 720,
+            scaleFactor = 720f / screenBounds.bottom
+        )
+    }
+
+    override fun register(hub: IHub, options: SentryOptions) {
+        this.options = options
+
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
+            options.logger.log(INFO, "Session replay is only supported on API 26 and above")
+            return
+        }
+
+        // TODO: check for replaysSessionSampleRate and replaysOnErrorSampleRate
+
+        this.hub = hub
+        recorder = WindowRecorder(options, recorderConfig, this)
+        isEnabled.set(true)
+    }
+
+    fun isRecording() = isRecording.get()
+
+    fun start() {
+        if (!isEnabled.get()) {
+            options.logger.log(
+                DEBUG,
+                "Session replay is disabled due to conditions not met in Integration.register"
+            )
+            return
+        }
+
+        if (isRecording.getAndSet(true)) {
+            options.logger.log(
+                DEBUG,
+                "Session replay is already being recorded, not starting a new one"
+            )
+            return
+        }
+
+        currentSegment.set(0)
+        currentReplayId.set(SentryId())
+        hub?.configureScope { it.replayId = currentReplayId.get() }
+        cache = ReplayCache(options, currentReplayId.get(), recorderConfig)
+
+        recorder?.startRecording()
+        segmentTimestamp.set(DateUtils.getCurrentDateTime())
+        // TODO: finalize old recording if there's some left on disk and send it using the replayId from persisted scope (e.g. for ANRs)
+    }
+
+    fun resume() {
+        segmentTimestamp.set(DateUtils.getCurrentDateTime())
+        recorder?.resume()
+    }
+
+    fun pause() {
+        val now = dateProvider.currentTimeMillis
```

Comment:
I guess we need to align the date provider usage to have the same source of truth, e.g. `DateUtils.getCurrentDateTime()` vs . `DateUtils.getCurrentDateTime()`

Reply:
> DateUtils.getCurrentDateTime() vs . DateUtils.getCurrentDateTime()

you mean `dateProvider.currentTimeMillis vs DateUtils.getCurrentDateTime()` right?

I'd love to align them, but the thing is that rrweb expects timestamp as `long` in milliseconds, and relay expects Event timestamp in ISO format, hence the divergence. I'll leave a TODO here though, to explore if we can align those 👍

Reply:
yeah for sure I'll change that to be able to test it, will do that later when adding tests 👍